dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/TrackerBundleParamsConverter.java

```diff
@@ -0,0 +1,182 @@
+ *
+ * Assuming a structure like:
+ * <pre>
+ *
+ * TrackerBundleParams
+ *   |
+ *   __TEI
+ *      |_ENROLLMENT 1
+ *      |      |
+ *      |      |_ EVENT 1
+ *      |      |
+ *      |      |_ EVENT 2
+ *      |
+ *      |_ENROLLMENT 2
+ *            |
+ *            |_ EVENT 3
+ *            |_ EVENT 4
+ * </pre>
+ *
+ * This converter will transform the object into:
+ *
+ * <pre>
+ *
+ * TrackerBundleParams
+ *  |
+ *  |___TEI
+ *  |___ENROLLMENT 1, ENROLLMENT 2
+ *  |
+ *  |___EVENT 1, EVENT 2, EVENT 3, EVENT 4
+ *
+ * </pre>
+ *
+ * This converter also assigns UIDs to Tracked Entities, Enrollment and Events if the payload does not contain UIDs
+ *
+ * @author Luciano Fiandesio
+ */
+public class TrackerBundleParamsConverter
+    extends
+    StdConverter<TrackerBundleParams, TrackerBundleParams>
+{
+    @Override
+    public TrackerBundleParams convert( TrackerBundleParams bundle )
+    {
+        if ( hasNestedStructure( bundle ) )
+        {
+            // pre assign UIDs to entities, if UIDs are missing
+            generateUid( bundle );
+
+            flattenPayload( bundle );
+        }
+
+        return bundle;
+    }
+
+    private void flattenPayload( TrackerBundleParams bundle ) {
+
+        List<Event> events = new ArrayList<>();
+
+        List<Enrollment> enrollments = new ArrayList<>();
+
+        for ( Enrollment enrollment : bundle.getTrackedEntities().stream()
+                .flatMap( l -> l.getEnrollments().stream() ).collect( Collectors.toList() ) )
+        {
+            events.addAll( enrollment.getEvents().stream().map( e -> addParent( e, enrollment.getEnrollment() ) )
+                .collect( Collectors.toList() ) );
+
+            // remove events from enrollment
+            enrollment.setEvents( Collections.emptyList() );
+        }
+
+        for ( TrackedEntity trackedEntity : bundle.getTrackedEntities() )
+        {
+            enrollments.addAll( trackedEntity.getEnrollments().stream()
+                .map( e -> addParent( e, trackedEntity.getTrackedEntity() ) ).collect( Collectors.toList() ) );
+
+            // remove enrollments from tracked entities
+            trackedEntity.setEnrollments( Collections.emptyList() );
+        }
+
+        bundle.getEvents().addAll( events );
+        bundle.getEnrollments().addAll( enrollments );
+    }
+
+    /**
+     * Make sure that the Enrollment has the parent ID correctly set
+     *
+     */
+    private Enrollment addParent( Enrollment enrollment, String trackerEntityId )
+    {
+        if ( StringUtils.isEmpty( enrollment.getTrackedEntity() ) )
+        {
+            enrollment.setTrackedEntity( trackerEntityId );
+        }
+        return enrollment;
+    }
+
+    /**
+     * Make sure that the Event has the parent ID correctly set
+     *
+     */
+    private Event addParent( Event event, String enrollmentId )
+    {
+
+        if ( StringUtils.isEmpty( event.getEnrollment() ) )
```

Comment:
Same here.. what if nested event is pointing to different enrollment... validation error

Reply:
Discussed with Morten, validation should not be taking place in this component